File: crates/hushd/src/siem/transforms/ocsf.rs
use crate::siem::types::{Outcome, ResourceType, SecurityEvent, SecuritySeverity};

/// Convert a `SecurityEvent` to a lightweight OCSF-like JSON payload.
///
/// Note: This is not a complete OCSF class mapping. It provides a stable, self-describing
/// structure that can be evolved toward full OCSF compliance.
pub fn to_ocsf(event: &SecurityEvent) -> serde_json::Value {
    let (severity_id, severity) = ocsf_severity(&event.decision.severity);
    let status = match event.outcome {
        Outcome::Success => "success",
        Outcome::Failure => "failure",
        Outcome::Unknown => "unknown",
    };

    let mut out = serde_json::json!({
        "time": event.timestamp_rfc3339_nanos(),
        "severity_id": severity_id,
        "severity": severity,
        "status": status,
        "activity_name": format!("{:?}", event.event_type),
        "category_name": format!("{:?}", event.event_category),
        "message": event.decision.reason,
        "metadata": {
            "version": event.schema_version,
            "product": {
                "name": "clawdstrike",
                "version": event.agent.version,
            }
        },
        "actor": {
            "id": event.agent.id,
            "name": event.agent.name,
            "type": event.agent.agent_type,
        },
        "session": {
            "id": event.session.id,
            "tenant_id": event.session.tenant_id,
            "environment": event.session.environment,
        },
        "decision": {
            "allowed": event.decision.allowed,
            "guard": event.decision.guard,
            "severity": format!("{:?}", event.decision.severity),
        },
    });

    match event.resource.resource_type {
        ResourceType::File => {
            out["file"] = serde_json::json!({
                "path": event.resource.path,
                "name": event.resource.name,
            });
        }
        ResourceType::Network => {
            out["network"] = serde_json::json!({
                "host": event.resource.host,
                "port": event.resource.port,
            });
        }
        ResourceType::Process => {
            out["process"] = serde_json::json!({
                "command_line": event.resource.name,
            });
        }
        ResourceType::Tool => {
            out["tool"] = serde_json::json!({
                "name": event.resource.name,
            });
        }
        ResourceType::Configuration => {}
    }

    out
}

fn ocsf_severity(sev: &SecuritySeverity) -> (u8, &'static str) {
    // OCSF severity_id is typically 1-6 (Informational..Critical). Keep a stable mapping.
    match sev {
        SecuritySeverity::Info => (1, "informational"),
        SecuritySeverity::Low => (2, "low"),
        SecuritySeverity::Medium => (3, "medium"),
        SecuritySeverity::High => (4, "high"),
        SecuritySeverity::Critical => (6, "critical"),
    }
}

File: packages/sdk/hush-go/siem/transforms/ocsf.go
package transforms

import (
	"github.com/backbay-labs/clawdstrike-go/siem"
)

// OCSF Activity IDs for Security Finding.
const (
	ocsfClassUID    = 2001 // Security Finding
	ocsfActivityNew = 1    // Create
)

// outcomeToOCSF maps Clawdstrike outcomes to OCSF status IDs.
var outcomeToOCSF = map[string]int{
	"allow": 1, // Success
	"deny":  2, // Failure
	"warn":  0, // Unknown
}

// severityToOCSF maps severity strings to OCSF severity IDs.
var severityToOCSF = map[string]int{
	"info":     1,
	"warning":  3,
	"warn":     3,
	"error":    4,
	"low":      2,
	"medium":   3,
	"high":     4,
	"critical": 5,
}

// ToOCSF converts a SecurityEvent to Open Cybersecurity Schema Framework format.
func ToOCSF(event siem.SecurityEvent) map[string]interface{} {
	statusID := 0
	if id, ok := outcomeToOCSF[event.Outcome]; ok {
		statusID = id
	}

	ocsf := map[string]interface{}{
		"class_uid":   ocsfClassUID,
		"activity_id": ocsfActivityNew,
		"time":        event.Timestamp.UnixMilli(),
		"status_id":   statusID,
		"message":     event.EventType,
		"metadata": map[string]interface{}{
			"version":      event.SchemaVersion,
			"product":      map[string]string{"name": "clawdstrike", "vendor_name": "backbay"},
			"original_uid": event.EventID,
		},
		"actor": map[string]interface{}{
			"agent": map[string]interface{}{
				"uid":       event.Agent.ID,
				"name":      event.Agent.Name,
				"type_name": event.Agent.Type,
			},
		},
	}

	if event.Decision != nil {
		sevID := 0
		if id, ok := severityToOCSF[event.Decision.Severity]; ok {
			sevID = id
		}
		ocsf["severity_id"] = sevID
		ocsf["finding"] = map[string]interface{}{
			"title":   event.Decision.Guard,
			"message": event.Decision.Message,
		}
	}

	if event.Resource != nil {
		resources := []map[string]interface{}{
			{
				"type": event.Resource.Type,
				"name": resourceName(event.Resource),
			},
		}
		ocsf["resources"] = resources
	}

	if event.Threat != nil {
		ocsf["confidence_id"] = int(event.Threat.Confidence * 10)
		ocsf["analytic"] = map[string]interface{}{
			"category": event.Threat.Category,
		}
	}

	return ocsf
}

func resourceName(r *siem.ResourceInfo) string {
	switch r.Type {
	case "file":
		return r.Path
	case "network":
		return r.Host
	case "tool":
		return r.Tool
	default:
		return ""
	}
}
